sql/plan/load_data.go

```diff
@@ -0,0 +1,379 @@
+		fileName = Tmpfiledir + TmpfileName
+	}
+
+	file, err := os.Open(fileName)
+	if err != nil {
+		return nil, err
+	}
+
+	scanner := bufio.NewScanner(file)
+
+	// Set the split function for lines.
+	l.parseLines(scanner)
+
+	// Skip through the lines that need to be ignored.
+	for l.IgnoreNum > 0 && scanner.Scan() {
+		scanner.Text()
+		l.IgnoreNum--
+	}
+
+	return &loadDataIter{
+		scanner:                 scanner,
+		destination:             l.Destination,
+		ctx:                     ctx,
+		file:                    file,
+		local:                   l.Local,
+		fieldsTerminatedByDelim: l.fieldsTerminatedByDelim,
+		fieldsEnclosedByDelim:   l.fieldsEnclosedByDelim,
+		fieldsOptionallyDelim:   l.fieldsOptionallyDelim,
+		linesTerminatedByDelim:  l.linesTerminatedByDelim,
+		linesStartingByDelim:    l.linesStartingByDelim,
+	}, nil
+}
+
+// parseLines finds the delim that terminates each line and returns the overall line.
+func (l LoadData) parseLines(scanner *bufio.Scanner) {
+	splitFunc := func(data []byte, atEOF bool) (advance int, token []byte, err error) {
+		// Return nothing if at end of file and no data passed.
+		if atEOF && len(data) == 0 {
+			return 0, nil, nil
+		}
+
+		// Find the index of the LINES TERMINATED BY delim.
+		if i := strings.Index(string(data), l.linesTerminatedByDelim); i >= 0 {
+			return i + 1, data[0:i], nil
+		}
+
+		// If at end of file with data return the data.
+		if atEOF {
+			return len(data), data, nil
+		}
+
+		return
+	}
+	scanner.Split(splitFunc)
+}
+
+type loadDataIter struct {
+	scanner                 *bufio.Scanner
+	destination             sql.Node
+	ctx                     *sql.Context
+	file                    *os.File
+	local                   bool
+	fieldsTerminatedByDelim string
+	fieldsEnclosedByDelim   string
+	fieldsOptionallyDelim   bool
+	fieldsEscapedByDelim    string
+	linesTerminatedByDelim  string
+	linesStartingByDelim    string
+}
+
+func (l loadDataIter) Next() (returnRow sql.Row, returnErr error) {
+	keepGoing := l.scanner.Scan()
+	if !keepGoing {
+		return nil, io.EOF
+	}
+
+	line := l.scanner.Text()
+	exprs, err := l.parseFields(line)
+
+	if err != nil {
+		return nil, err
+	}
+
+	// If exprs is nil then this is a skipped line (see test cases). Keep skipping
+	// until exprs != nil
+	for exprs == nil {
+		keepGoing = l.scanner.Scan()
+		if !keepGoing {
+			return nil, io.EOF
+		}
+
+		line = l.scanner.Text()
+		exprs, err = l.parseFields(line)
+
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// Match input columns with the amount of columns provided in the text.
+	// Append nils to the parsed fields if they are less than the input columns.
+	// TODO: Match schema with column order
+	colDiff := len(l.destination.Schema()) - len(exprs)
+
+	// append NULLS for the rest of the fields
+	exprs = addNullsToValues(exprs, colDiff)
+
+	// create the values that are returned as a row iter.
+	var values [][]sql.Expression
+	values = append(values, exprs)
+	newValue := NewValues(values)
```

Comment:
This creates a lot of garbage just to return a sql.Row. Just return the row directly